osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs

```diff
@@ -77,6 +81,19 @@ public override void ApplyBeatmap(IBeatmap beatmap)
             if (beatmap.HitObjects.Count > 0)
                 gameplayEndTime = beatmap.HitObjects[^1].GetEndTime();
 
+            noDrainPeriodTracker = new PeriodTracker(beatmap.Breaks.Select(breakPeriod => new Period(
+                beatmap.HitObjects
+                       .Select(hitObject => hitObject.GetEndTime())
+                       .Where(endTime => endTime < breakPeriod.StartTime)
+                       .DefaultIfEmpty(double.MinValue)
+                       .Last(),
+                beatmap.HitObjects
+                       .Select(hitObject => hitObject.StartTime)
+                       .Where(startTime => startTime > breakPeriod.EndTime)
```

Comment:
What happens if the hitobject's start time is equal to the break's end time? E.g.

```
[H] - - [B - - - [HB] - [H]
  ^^^^^^^^^^^^^^^^^^^^^^^
  ^^^^^^^^^^^^^^^^   
```
Won't the drain region be the first one marked above, rather than the second one? That is to say, these checks should probably be `>=` and `<=` above.

This also maybe shows a little bit of an interesting scenario scenario:

```
[H] - - [B - [H] - B] - [H]
  ^^^^^^^^^^^^ ^^^^^^^^^^
  ^^^^^^^^^^^^^^^^^^^^^^^
```
Where if a hitobject intersects a break, I'd expect the no-drain region to be the first one marked above but it seems like this code will do the second one? It's a bit of an edge-case scenario and I only know of aspire maps that abuse this, so maybe not worth consideration.

Reply:
Good points. I'll add these as test cases and see if changing the comparison suffices.

Reply:
I think this second case is fine to disregard for now.